dcap-provider/src/lib.rs

```diff
@@ -0,0 +1,373 @@
+// TCB”, meaning `PPID_RSA3072_ENCRYPTED`. Hopefully, the QL will call
+// `sgx_ql_get_quote_config` again in the same instance of this module, except
+// now `get_certdata` will return early with an error because it is
+// re-entered. This error should cause the QL to generate a
+// `PPID_RSA3072_ENCRYPTED` quote.
+//
+// It doesn't matter which enclave the quote is for, so we just use a
+// purpose-built built-in enclave to get a report.
+fn get_quote_with_raw_tcb() -> Result<Vec<u8>, Quote3Error> {
+    let ti = dcap_ql::target_info().map_err(|e| {
+        error!("PPID query: failed to obtain target info: {:?}", e);
+        Quote3Error::NoPlatformCertData
+    })?;
+    let mut loader = dcap_ql::enclave_loader().map_err(|e| {
+        error!("PPID query: failed to load enclave loader: {}", e);
+        Quote3Error::NoPlatformCertData
+    })?;
+    let report = report_test::report(&ti, &mut loader).map_err(|e| {
+        error!("PPID query: {}", e);
+        Quote3Error::NoPlatformCertData
+    })?;
+    let quote = dcap_ql::quote(&report).map_err(|e| {
+        error!("PPID query: failed to obtain quote: {:?}", e);
+        Quote3Error::NoPlatformCertData
+    })?;
+    Ok(quote)
+}
+
+/// Check if a quote matches the QE ID and parse the QE certification data from
+/// the quote.
+fn parse_certdata(
+    qe_id: quote::QeId,
+    quote: &[u8],
+) -> Result<quote::Qe3CertDataPpid<'static>, Quote3Error> {
+    let quote = quote::Quote::parse(quote).map_err(|e| {
+        error!("PPID query: failed to parse quote: {}", e);
+        Quote3Error::NoPlatformCertData
+    })?;
+
+    let &quote::QuoteHeader::V3 {
+        ref qe3_vendor_id,
+        ref user_data,
+        ..
+    } = quote.header();
+
+    if qe3_vendor_id != &&quote::QE3_VENDOR_ID_INTEL[..] {
+        error!("PPID query: QE vendor ID is not Intel");
+        return Err(Quote3Error::NoPlatformCertData);
+    }
+
+    let mut qe_id = qe_id.into_owned();
+    qe_id.resize(20usize, 0u8);
+    if qe_id != &user_data[..] {
+        error!("PPID query: QE ID mismatch");
+        return Err(Quote3Error::NoPlatformCertData);
+    }
+
+    let sig = quote
+        .signature::<quote::Quote3SignatureEcdsaP256>()
+        .map_err(|e| {
+            error!("PPID query: {}", e);
+            Quote3Error::NoPlatformCertData
+        })?;
+
+    match sig.certification_data::<quote::Qe3CertDataPpid>() {
+        Ok(_)
+            if sig.certification_data_type()
+                != quote::CertificationDataType::PpidEncryptedRsa3072 =>
+        {
+            error!(
+                "PPID query: Invalid certification data type: {:?}",
+                sig.certification_data_type()
+            );
+            Err(Quote3Error::NoPlatformCertData)
+        }
+        Err(e) => {
+            error!("PPID query: {}", e);
+            Err(Quote3Error::NoPlatformCertData)
+        }
+        Ok(cd) => Ok(cd.clone_owned()),
+    }
```

Comment:
> Are we not checking sig.certification_data_type() before sig.certification_data::<quote::Qe3CertDataPpid>()

`sig.certification_data::<quote::Qe3CertDataPpid>()` *already* checks `sig.certification_data_type()`. However, it will deserialize types other than `PpidEncryptedRsa3072`, which are not supported here.

I'll rewrite as suggested.